### PR TITLE
SP-8416: implement timestamps for application retransmissions

### PR DESCRIPTION
**Pull request by przemyslawfierek-silvair**

### CODE REVIEW

File: mesh/node.c [hunk 1]
```diff
@@ -222,7 +224,35 @@ static void set_defaults(struct mesh_node *node)
 	node->fd_io = NULL;
 }
 
-static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
+struct postponed_msg {
+	struct fd_msg *msg;
+	size_t msg_len;
+};
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:230]:
```c
struct postponed_msg {
    size_t msg_len;
    struct fd_msg fd_msg;
}

static struct postponed_msg *postponed_msg_new(struct fd_msg *fd_msg, size_t msg_len)
{
    struct postponed_msg *msg = l_malloc(sizeof(*msg) + msg_len);
    msg->msg_len = msg_len;
    memcpy(&msg->fd_msg, fd_msg, msg_len);
    return msg;
}
```
avoids the 2nd `l_malloc` and you don't need `postponed_msg_free` anymore

File: mesh/node.c [hunk 2]
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
 			msg->element_idx, msg->dst_addr, msg->ttl, data_len);
 
 	if (!mesh_model_send(node, src, msg->dst_addr, app_idx, msg->net_idx,
-		      msg->ttl, FD_MSG_IS_SEGMENTED(msg), data_len, msg->data))
+		msg->ttl, FD_MSG_IS_SEGMENTED(msg), data_len, msg->data))
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:273]: I think this should be indented (but with a tab). Separate commit?

File: mesh/node.c [hunk 3]
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
...
+	struct mesh_node *node = user_data;
+	uint64_t current_timestamp = get_timestamp_ms();
+	
+	struct postponed_msg *postponed_msg =
+				l_queue_peek_head(node->rc_priority_queue);
+
+	if (!postponed_msg)
+		return;
+
+	if (postponed_msg->msg->timestamp <= current_timestamp) {
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:308]: This should be a `while`

File: mesh/node.c [hunk 4]
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
...
+	if (postponed_msg->msg->timestamp > current_timestamp) {
+		uint64_t diff =
+			postponed_msg->msg->timestamp - current_timestamp;
+		l_timeout_modify_ms(timeout, diff ?: 1);
+	}
+	else {
+		/* to avoid recursive function call, schedule another timeout */
+		l_timeout_modify_ms(timeout, 1);
+	}
+}
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:306]: If you process messages in a loop, you don't need to worry about recurstion:

```c
static void try_send_rc_msg(struct l_timeout *timeout, void *user_data)
{
    struct mesh_node *node = user_data;
    uint64_t current_timestamp = get_timestamp_ms();
    struct postponed_msg *postponed_msg;

    while ((postponed_msg = l_queue_peek_head(node->rc_priority_queue))) {
        if (postponed_msg->msg->timestamp > current_timestamp)
            break;

        send_rc_msg(postponed_msg->msg, postponed_msg->msg_len, node);
        l_queue_pop_head(node->rc_priority_queue);
        l_free(postponed_msg);
    }

    if (postponed_msg) {
        uint64_t diff = postponed_msg->msg->timestamp - current_timestamp;
        l_timeout_modify_ms(timeout, diff);
    }
}
```

File: mesh/node.c [hunk 5]
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
...
 
+static int compare_msgs_timestamp(const void *a, const void *b, void *user_data)
+{
+	const struct postponed_msg *msg1 = a;
+	const struct postponed_msg *msg2 = b;
+
+	return msg2->msg->timestamp >= msg1->msg->timestamp;
+}
+
+static void try_send_rc_msg(struct l_timeout *timeout, void *user_data)
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: I'd include "flush" somewhere in the function name ;)

File: mesh/node.c [hunk 6]
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
...
+			postponed_msg->msg->timestamp - current_timestamp;
+		l_timeout_modify_ms(timeout, diff ?: 1);
+	}
+	else {
+		/* to avoid recursive function call, schedule another timeout */
+		l_timeout_modify_ms(timeout, 1);
+	}
+}
+
+static void rc_msg_postpone_rc_msg(struct fd_msg *msg, size_t msg_len,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: `rc_msg_postpone` would be enough, or maybe just `rc_msg_push`

File: mesh/util.c [hunk 7]
```diff
@@ -50,6 +50,14 @@ void print_packet(const char *label, const void *data, uint16_t size)
 				(uint32_t) pkt_time.tv_usec/1000, label);
 }
 
+uint64_t get_timestamp_ms(void)
```
Comment (michallowasrzechonek-silvair) [inline mesh/util.c:53]: See `l_time_*` and friends.

File: mesh/model.c [hunk 8]
```diff
@@ -802,15 +802,6 @@ static inline void send_fd(void *data, size_t len, void *user_data)
 	fd_msg_send(user_data, data, len);
 }
 
-static inline uint64_t get_timestamp(void)
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c:805]: Separate commit

File: mesh/util.c (hunk 7 above)
Comment (przemyslawfierek-silvair) [reply mesh/util.c:53]: `l_time_now` returns time since device boot ;/

**Review by michallowasrzechonek-silvair — APPROVED** [review]

**Comment by michallowasrzechonek-silvair** [comment]
Do we have tests?

**Comment by przemyslawfierek-silvair** [comment]
Not yet

**Pull request merged**